cpmpy/solvers/gurobi.py

```diff
@@ -0,0 +1,376 @@
+
+        cpm_cons = flatten_constraint(cpm_con)
+        cpm_cons = linearize_constraint(cpm_cons)
+
+        for con in cpm_cons:
+            self._post_constraint(con)
+
+        return self
+
+    def solve(self, time_limit=None, solution_callback=None, **kwargs):
+        """
+            Call the gurobi solver
+
+            Arguments:
+            - time_limit:  maximum solve time in seconds (float, optional)
+            - kwargs:      any keyword argument, sets parameters of solver object
+
+            Arguments that correspond to solver parameters:
+            <Please document key solver arguments that the user might wish to change
+             for example: log_output=True, var_ordering=3, num_cores=8, ...>
+            <Add link to documentation of all solver parameters>
+        """
+        from gurobipy import GRB
+
+        if time_limit is not None:
+            raise NotImplementedError("TEMPLATE: TODO, implement time_limit")
+
+        # call the solver, with parameters
+        for param, val in kwargs.items():
+            self.gbi_model.setParam(param, val)
+
+        _ = self.gbi_model.optimize(callback=solution_callback)
+
+        my_status = self.gbi_model.Status
+
+        # new status, translate runtime
+        self.cpm_status = SolverStatus(self.name)
+        self.cpm_status.runtime = self.gbi_model.runtime
+
+        # translate exit status
+        if my_status == GRB.OPTIMAL and self._objective_value is None:
+            self.cpm_status.exitstatus = ExitStatus.FEASIBLE
+        elif my_status == GRB.OPTIMAL and self._objective_value is not None:
+            self.cpm_status.exitstatus = ExitStatus.OPTIMAL
+        elif my_status == GRB.INFEASIBLE:
+            self.cpm_status.exitstatus = ExitStatus.UNSATISFIABLE
+        else:  # another?
+            raise NotImplementedError(
+                f"Translation of gurobi status {my_status} to CPMpy status not implemented")  # a new status type was introduced, please report on github
+
+        # True/False depending on self.cpm_status
+        has_sol = self._solve_return(self.cpm_status)
+
+        # translate solution values (of user vars only)
+        if has_sol:
+            # fill in variable values
+            for cpm_var in self.user_vars:
+                solver_val = self.solver_var(cpm_var).X
+                if cpm_var.is_bool():
+                    cpm_var._value = bool(solver_val)
+                else:
+                    cpm_var._value = int(solver_val)
+
+        return has_sol
+
+    def solver_var(self, cpm_var):
+        """
+            Creates solver variable for cpmpy variable
+            or returns from cache if previously created
+        """
+        from gurobipy import GRB
+
+        if is_num(cpm_var):
+            return cpm_var
+
+        # special case, negative-bool-view
+        # work directly on var inside the view
+        if isinstance(cpm_var, NegBoolView):
+            return 1-self.solver_var(cpm_var._bv)
```

Comment:
ohhh, this worries me

solver_var() should return a solver-native object, not an expression...

This can break a lot of things, e.g. or(~bv0, ~bv1, ~bv2)?

Is there a gurobi-native way to do negation?

If not, maybe we should add a transformation that replaces ~bv by 1-bv in all expressions, and only continue from that?

Reply:
After doing some more research, I don't think there is native gurobi way of doing this. I'll add a transformation for it then